File: src/main/java/com/sandjelkovic/dispatchd/provider/TraktMediaProvider.java
package com.sandjelkovic.dispatchd.provider;

import com.sandjelkovic.dispatchd.trakt.dto.EpisodeTrakt;
import com.sandjelkovic.dispatchd.trakt.dto.SeasonTrakt;
import com.sandjelkovic.dispatchd.trakt.dto.ShowUpdateTrakt;
import com.sandjelkovic.dispatchd.trakt.dto.TvShowTrakt;
import org.springframework.scheduling.annotation.Async;

import java.time.LocalDate;
import java.util.List;
import java.util.concurrent.Future;

public interface TraktMediaProvider {

	TvShowTrakt getTvShow(String showId);

	EpisodeTrakt getEpisode(String showId, String seasonNumber, String episodeNumber);

	List<EpisodeTrakt> getSeasonEpisodes(String showId, String seasonNumber);

	List<EpisodeTrakt> getShowEpisodes(String showId);

	List<SeasonTrakt> getSeasons(String showId);

	List<SeasonTrakt> getSeasonsMinimal(String showId);

	SeasonTrakt getSeason(String showId, String seasonNumber);

	List<ShowUpdateTrakt> getUpdates(LocalDate fromDate);

	@Async
	Future<TvShowTrakt> getTvShowAsync(String showId);

	@Async
	Future<EpisodeTrakt> getEpisodeAsync(String showId, String seasonNumber, String episodeNumber);

	@Async
	Future<List<EpisodeTrakt>> getSeasonEpisodesAsync(String showId, String seasonNumber);

	@Async
	Future<List<EpisodeTrakt>> getShowEpisodesAsync(String showId);

	@Async
	Future<List<SeasonTrakt>> getSeasonsAsync(String showId);

	@Async
	Future<List<SeasonTrakt>> getSeasonsMinimalAsync(String showId);

	@Async
	Future<SeasonTrakt> getSeasonAsync(String showId, String seasonNumber);

	@Async
	Future<List<ShowUpdateTrakt>> getUpdatesAsync(LocalDate fromDate);

}

File: content-service/src/main/kotlin/com/sandjelkovic/dispatchd/content/trakt/provider/TraktMediaProvider.kt
package com.sandjelkovic.dispatchd.content.trakt.provider

import arrow.core.Option
import arrow.core.Try
import com.sandjelkovic.dispatchd.content.trakt.dto.EpisodeTrakt
import com.sandjelkovic.dispatchd.content.trakt.dto.SeasonTrakt
import com.sandjelkovic.dispatchd.content.trakt.dto.ShowTrakt
import com.sandjelkovic.dispatchd.content.trakt.dto.ShowUpdateTrakt
import org.springframework.scheduling.annotation.Async
import org.springframework.scheduling.annotation.AsyncResult
import java.time.LocalDate

/**
 * @author sandjelkovic
 * @date 28.1.18.
 */
interface TraktMediaProvider {
    fun getShow(showId: String): Try<Option<ShowTrakt>>
    fun getShowEpisodes(showId: String): Try<List<EpisodeTrakt>>
    fun getSeasons(showId: String): Try<List<SeasonTrakt>>
    fun getSeasonsMinimal(showId: String): Try<List<SeasonTrakt>>
    fun getSeasonEpisodes(showId: String, seasonNumber: String): Try<List<EpisodeTrakt>>
    fun getUpdates(fromDate: LocalDate): Try<List<ShowUpdateTrakt>>

    //TODO convert to coroutines or Arrow.IO
    @Async
    fun getSeasonsAsync(showId: String): AsyncResult<Try<List<SeasonTrakt>>>

    @Async
    fun getShowEpisodesAsync(showId: String): AsyncResult<Try<List<EpisodeTrakt>>>
}
